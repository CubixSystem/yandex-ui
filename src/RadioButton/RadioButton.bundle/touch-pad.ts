import { FC } from 'react';
import { compose, composeU } from '@bem-react/core';

import {
    IRadioButtonProps as IRadioButtonTouchPadProps,
    RadioButton as RadioButtonTouchPad,
} from '../RadioButton@touch-pad';
// _size
import { withSizeS } from '../_size/RadioButton_size_s';
import { withSizeM } from '../_size/RadioButton_size_m';
import { withSizeL } from '../_size/RadioButton_size_l';
// _view
import { withViewDefault } from '../_view/RadioButton_view_default';

export * from '../RadioButton@touch-pad';

export interface IRadioButtonProps extends IRadioButtonTouchPadProps {
    size?: 's' | 'm' | 'l';
    view?: 'default';
}

export const RadioButton = compose(
    composeU(withSizeS, withSizeM, withSizeL),
    withViewDefault,
)(RadioButtonTouchPad) as FC<IRadioButtonProps>;
