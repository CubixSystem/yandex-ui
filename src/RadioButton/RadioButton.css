.RadioButton {
    position: relative;

    display: inline-flex;

    font-family: var(--radio-button-font-family);
    user-select: none;
}
