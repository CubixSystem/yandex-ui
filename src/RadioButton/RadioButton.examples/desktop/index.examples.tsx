import { withKnobs } from '@storybook/addon-knobs';

import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export * from './view.examples';
export * from './size.examples';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: [withKnobs, ...createDecorators({ scope: 'desktop' })],
    parameters,
};
