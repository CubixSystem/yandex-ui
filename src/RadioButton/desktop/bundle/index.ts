export * from '../../RadioButton.bundle/desktop';
