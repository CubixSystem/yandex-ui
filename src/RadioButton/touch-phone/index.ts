export * from '../RadioButton@touch-phone';
export * from '../_size/RadioButton_size_l';
export * from '../_size/RadioButton_size_m';
export * from '../_size/RadioButton_size_s';
export * from '../_view/RadioButton_view_default';
