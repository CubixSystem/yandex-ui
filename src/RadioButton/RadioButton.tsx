import React, { ReactNode, RefObject, CSSProperties, useRef, useCallback, FC } from 'react';
import { useComponentRegistry } from '@bem-react/di';
import { cn } from '@bem-react/classname';

import { useUniqId } from '../useUniqId';
import { Omit } from '../typings/utility-types';
import { IRadioButtonRegistry } from './RadioButton.registry';
import { IRadioButtonControlProps } from './Control/RadioButton-Control';
import './RadioButton.css';

export const cnRadioButton = cn('RadioButton');

export interface IRadioButtonOption extends Omit<IRadioButtonControlProps, 'name' | 'checked'> {
    /**
     * Содержимое кнопки.
     */
    children: ReactNode;
}

export interface IRadioButtonProps extends Omit<IRadioButtonControlProps, 'id' | 'checked' | 'controlRef'> {
    /**
     * Набор опций.
     */
    options: IRadioButtonOption[];

    /**
     * Ссылка на корневой DOM-элемент компонента.
     */
    innerRef?: RefObject<HTMLSpanElement>;

    /**
     * Пользовательские стили на корневом DOM-элементе.
     */
    style?: CSSProperties;

    /**
     * Метка для радиогруппы.
     */
    'aria-label'?: string;
}

/**
 * Компонент для создания радиогруппы.
 * @param {IRadioButtonProps} props
 */
export const RadioButton: FC<IRadioButtonProps> = ({
    className,
    style,
    innerRef,
    'aria-label': ariaLabel,
    options,
    value: rootValue,
    disabled: rootDisabled,
    // eslint-disable-next-line react-hooks/rules-of-hooks
    name = useUniqId(),
    // @ts-ignore
    size: _size,
    // @ts-ignore
    view: _view,
    ...props
}) => {
    const { Plate, Radio, Control, Content } = useComponentRegistry<IRadioButtonRegistry>(cnRadioButton());

    const plateRef = useRef<HTMLDivElement>(null);
    const radioRef = useRef<HTMLLabelElement>();
    const radioCheckedRef = useCallback((radioCheckedNode) => {
        if (!radioCheckedNode) return;

        const plateNode = plateRef.current;

        if (!plateNode) return;

        const radioUncheckedNode = radioRef.current;

        // Показываем Plate только во время перехода между двумя Radio
        if (radioUncheckedNode && radioUncheckedNode !== radioCheckedNode) {
            const setPlateStyle = (node: HTMLElement) => {
                plateNode.style.left = `${node.offsetLeft}px`;
                plateNode.style.width = `${node.offsetWidth}px`;
            };

            setPlateStyle(radioUncheckedNode);

            plateNode.hidden = false;

            setPlateStyle(radioCheckedNode);
        }

        radioRef.current = radioCheckedNode;
    }, []);

    return (
        <span
            className={cnRadioButton(null, [className])}
            style={style}
            ref={innerRef}
            role="radiogroup"
            aria-label={ariaLabel}
            onFocus={(event) => event.currentTarget.setAttribute('data-focused', '')}
            onBlur={(event) => event.currentTarget.removeAttribute('data-focused')}
        >
            <Plate innerRef={plateRef} onTransitionEnd={(event) => (event.currentTarget.hidden = true)} />
            {options.map(({ children, ...controlProps }) => {
                const checked = rootValue === controlProps.value;

                controlProps.disabled = rootDisabled || controlProps.disabled;

                return (
                    <Radio
                        checked={checked}
                        disabled={controlProps.disabled}
                        key={controlProps.value}
                        innerRef={checked ? radioCheckedRef : undefined}
                    >
                        <Control {...props} {...controlProps} checked={checked} name={name} />
                        <Content>{children}</Content>
                    </Radio>
                );
            })}
        </span>
    );
};

RadioButton.displayName = cnRadioButton();
