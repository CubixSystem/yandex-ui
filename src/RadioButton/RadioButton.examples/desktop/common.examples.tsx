import React, { useState } from 'react';

import { RadioButton } from '../../RadioButton.bundle/desktop';
import { EXAMPLE_DESKTOP_TOKEN, createDecorators, parameters } from '../examples-config';

export default {
    title: EXAMPLE_DESKTOP_TOKEN,
    decorators: createDecorators({ scope: 'desktop' }),
    parameters,
};

export const View = () => {
    const [value, setValue] = useState('a');

    const options = [
        { value: 'a', children: 'Option A' },
        { value: 'b', children: 'Option B' },
        { value: 'c', children: 'Option C' },
    ];

    return (
        <RadioButton
            size="m"
            view="default"
            value={value}
            options={options}
            onChange={(event) => setValue(event.target.value)}
        />
    );
};

View.story = {
    name: 'view',
};

export const Size = () => {
    const [valueS, setValueS] = useState('a');
    const [valueM, setValueM] = useState('a');
    const [valueL, setValueL] = useState('a');

    const options = [
        { value: 'a', children: 'Option A' },
        { value: 'b', children: 'Option B' },
        { value: 'c', children: 'Option C' },
    ];

    return (
        <>
            <RadioButton
                size="s"
                view="default"
                value={valueS}
                options={options}
                onChange={(event) => setValueS(event.target.value)}
            />
            <br />
            <br />
            <RadioButton
                size="m"
                view="default"
                value={valueM}
                options={options}
                onChange={(event) => setValueM(event.target.value)}
            />
            <br />
            <br />
            <RadioButton
                size="l"
                view="default"
                value={valueL}
                options={options}
                onChange={(event) => setValueL(event.target.value)}
            />
        </>
    );
};

Size.story = {
    name: 'size',
};
