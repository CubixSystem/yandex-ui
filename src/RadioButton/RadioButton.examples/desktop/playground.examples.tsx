import React, { useState } from 'react';
import { select, boolean, object } from '@storybook/addon-knobs';

import { cnTheme } from '../../../Theme';
import { presets, presetsKeys } from '../../../Theme/presets';
import { RadioButton } from '../../RadioButton.bundle/desktop';
import { Icon } from '../../../Icon/Icon.bundle/desktop';

export const Playground = () => {
    const [value, setValue] = useState('b');

    const preset = select('theme-preset', presetsKeys, 'default');
    const size = select('size', ['s', 'm', 'l'], 'm') as any;
    const view = select('view', ['default', ''], 'default') as any;
    const disabled = boolean('disabled', false);
    const options = object('options', [
        {
            value: 'a',
            children: 'Option A',
        },
        {
            value: 'b',
            children: (
                <Icon type="filter" style={{ width: 12 }} />
            ),
        },
        {
            value: 'c',
            children: (
                <>
                    <Icon glyph="type-check" style={{ width: 16, marginRight: 4 }} />
                    Option C
                </>
            ),
        },
        {
            value: 'd',
            children: (
                <>
                    Option D (
                    <b>disabled</b>
                    )
                </>
            ),
            disabled: true,
        },
        {
            value: 'e',
            children: (
                <>
                    Option E
                    <Icon glyph="type-check" style={{ width: 16, marginLeft: 4 }} />
                </>
            ),
        },
    ]);

    return (
        <div className={cnTheme(presets[preset])}>
            <RadioButton
                size={size}
                view={view}
                disabled={disabled}
                options={options}
                value={value}
                onChange={(event) => setValue(event.target.value)}
                aria-label="Радиогруппа"
            />
        </div>
    );
};

Playground.story = {
    name: 'playground',
};
