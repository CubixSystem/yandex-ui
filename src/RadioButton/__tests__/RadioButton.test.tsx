import React, { createRef } from 'react';

import { createClientRender, fireEvent, screen } from '../../internal/testing';
import { Keys } from '../../lib/keyboard';
import { RadioButton as RadioButtonCommon } from '../RadioButton';
import { RadioButton as RadioButtonDesktop } from '../desktop';
import { RadioButton as RadioButtonTouchPad } from '../touch-pad';
import { RadioButton as RadioButtonTouchPhone } from '../touch-phone';

const platforms = [
    ['desktop', RadioButtonDesktop],
    ['touch-pad', RadioButtonTouchPad],
    ['touch-phone', RadioButtonTouchPhone],
] as const;

const options = [
    { value: 'a', children: 'Option A' },
    { value: 'b', children: <>Option B</> },
];

describe.each(platforms)('RadioButton@%s', (_platform, RadioButton) => {
    const render = createClientRender();

    test('должен вернуть полный шаблон компонента (snapshot)', () => {
        const { container } = render(
            <RadioButton name="rb" value="b" options={options} className="mix" onChange={() => null} />,
        );

        expect(container.firstChild).toMatchSnapshot();
    });

    test('должен быть установлен корректный displayName', () => {
        expect(RadioButtonCommon.displayName).toBe('RadioButton');
    });

    test('должен устанавливать ссылку на корневой DOM элемент', () => {
        const innerRef = createRef<HTMLElement>();
        render(<RadioButton value="" options={[]} innerRef={innerRef} onChange={() => null} />);

        expect(innerRef.current).not.toBe(null);
    });

    test('должен вызывать событие onChange с event при клике на контрол', () => {
        // TODO: при переходе на 17 реакт `e.persist` убрать
        const onChange = jest.fn((e) => e.persist());
        render(<RadioButton value="b" options={options} onChange={onChange} />);

        fireEvent.click(screen.getAllByRole('radio')[0]);

        expect(onChange).toHaveBeenCalledTimes(1);

        expect(onChange.mock.calls[0][0].target.value).toBe('a');
    });

    test('при длительном зажатии клавиши отменяется стандартное поведение', () => {
        const preventDefault = jest.spyOn(KeyboardEvent.prototype, 'preventDefault');
        render(<RadioButton value="b" options={options} onChange={() => null} />);

        fireEvent.keyDown(screen.getByRole('radiogroup'), { keyCode: Keys.UP });

        expect(preventDefault).toHaveBeenCalledTimes(0);

        fireEvent.keyDown(screen.getByRole('radiogroup'), { keyCode: Keys.UP, repeat: true });

        expect(preventDefault).toHaveBeenCalledTimes(1);

        preventDefault.mockReset();
    });

    test('должен добавлять и удалять атрибут disabled при обновлении через props', () => {
        const { setProps } = render(<RadioButton value="b" options={options} onChange={() => null} />);

        expect(screen.getAllByRole('radio')[0]).not.toBeDisabled();
        expect(screen.getAllByRole('radio')[1]).not.toBeDisabled();

        setProps({ disabled: true });

        expect(screen.getAllByRole('radio')[0]).toBeDisabled();
        expect(screen.getAllByRole('radio')[1]).toBeDisabled();
    });

    test('должен позволять блокировать и разблокировать отдельные опции через props', () => {
        const { setProps } = render(<RadioButton value="b" options={options} onChange={() => null} />);

        expect(screen.getAllByRole('radio')[0]).not.toBeDisabled();
        expect(screen.getAllByRole('radio')[1]).not.toBeDisabled();

        const optionsWithBDisabled = [
            { value: 'a', label: 'Option A' },
            { value: 'b', label: 'Option B', disabled: true },
        ];

        setProps({ options: optionsWithBDisabled });

        expect(screen.getAllByRole('radio')[0]).not.toBeDisabled();
        expect(screen.getAllByRole('radio')[1]).toBeDisabled();
    });
});
