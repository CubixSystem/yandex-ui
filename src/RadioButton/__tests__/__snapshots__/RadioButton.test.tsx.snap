// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RadioButton@desktop должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="RadioButton mix"
  role="radiogroup"
>
  <div
    class="RadioButton-Plate"
    hidden=""
  />
  <label
    class="RadioButton-Radio"
  >
    <input
      autocomplete="off"
      class="RadioButton-Control"
      name="rb"
      type="radio"
      value="a"
    />
    <div
      class="RadioButton-Content"
    >
      Option A
    </div>
  </label>
  <label
    class="RadioButton-Radio RadioButton-Radio_checked"
  >
    <input
      autocomplete="off"
      checked=""
      class="RadioButton-Control"
      name="rb"
      type="radio"
      value="b"
    />
    <div
      class="RadioButton-Content"
    >
      Option B
    </div>
  </label>
</span>
`;

exports[`RadioButton@touch-pad должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="RadioButton mix"
  role="radiogroup"
>
  <div
    class="RadioButton-Plate"
    hidden=""
  />
  <label
    class="RadioButton-Radio"
  >
    <input
      autocomplete="off"
      class="RadioButton-Control"
      name="rb"
      type="radio"
      value="a"
    />
    <div
      class="RadioButton-Content"
    >
      Option A
    </div>
  </label>
  <label
    class="RadioButton-Radio RadioButton-Radio_checked"
  >
    <input
      autocomplete="off"
      checked=""
      class="RadioButton-Control"
      name="rb"
      type="radio"
      value="b"
    />
    <div
      class="RadioButton-Content"
    >
      Option B
    </div>
  </label>
</span>
`;

exports[`RadioButton@touch-phone должен вернуть полный шаблон компонента (snapshot) 1`] = `
<span
  class="RadioButton mix"
  role="radiogroup"
>
  <div
    class="RadioButton-Plate"
    hidden=""
  />
  <label
    class="RadioButton-Radio"
  >
    <input
      autocomplete="off"
      class="RadioButton-Control"
      name="rb"
      type="radio"
      value="a"
    />
    <div
      class="RadioButton-Content"
    >
      Option A
    </div>
  </label>
  <label
    class="RadioButton-Radio RadioButton-Radio_checked"
  >
    <input
      autocomplete="off"
      checked=""
      class="RadioButton-Control"
      name="rb"
      type="radio"
      value="b"
    />
    <div
      class="RadioButton-Content"
    >
      Option B
    </div>
  </label>
</span>
`;
